In [1]:
import pandas as pd
import numpy as np
import requests
import os
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

# Gather

In [2]:
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
folder_name = 'image_predictions'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

url_list = ['https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv']

for url in url_list:
    response = requests.get(url)
    with open(os.path.join(folder_name , url.split('/')[-1]) , mode = 'wb') as file:
             file.write(response.content)

In [4]:
image_prediction = pd.read_csv('image_predictions/image-predictions.tsv' , sep = '\t')

### Additional Data via the Twitter API

In [5]:
# I used configparser for my API keys and secret.
# This code returns error if you do not input your API and secret info.

import configparser

config = configparser.ConfigParser()

config['consumer'] = {
    'key': 'KTXWZ3Ucy7o9xodDyHNM9expc',
    'secret': 'kspq1bKtllzCVCK8XIlEuINz7qAzOw9xufM0OT2PWGvxGwbJUw'
}
config['access'] = {
    'token': '1160345017220812800-vitA9IyHFSrHg8zJnK2MjwZWk4lAEI',
    'secret': '5CFzbBJw8NzpDOxOiGs79PnBURGGEqZWKqOb6dZuJZsnt',
}
with open('config.ini', 'w') as file:
    config.write(file)

In [6]:
# read my APi info from config.ini

consumer_key = config['consumer']['key']
consumer_secret = config['consumer']['secret']
access_token = config['access']['token']
access_secret = config['access']['secret']


auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth , wait_on_rate_limit=True)

In [7]:
tweet_ids = twitter_archive.tweet_id.values

# Use timer to calculate how long this code takes to run
start = timer()

# Create a text file called tweet_json.txt
with open ('tweet_json.txt' , 'w') as outfile:
    for tweet_id in tweet_ids:
        try:
            # Get tweet informations
            tweet_info = api.get_status(tweet_id, tweet_mode='extended')
            #Write in the text file
            json.dump(tweet_info._json , outfile)
            # Put a indent for better readiness
            outfile.write('\n')
        
        # Error tweets
        except tweepy.TweepError as e:
            # Print tweet ID and error code description
            print ("Tweet ID:{}{}".format(str(tweet_id),e.reason))

# Timer calculation            
end = timer()
print(end - start , "seconds")

In [8]:
tweepy_data = pd.read_json('tweet_json.txt' , lines = True)

# Assess

## twitter_archive

In [9]:
twitter_archive.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [10]:
# tweet id has wrong datatype
# timestamp has wrong datatype

twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [11]:
# some strange names
twitter_archive.name.value_counts()

None         745
a             55
Charlie       12
Cooper        11
Lucy          11
            ... 
Apollo         1
Murphy         1
Sobe           1
Frönq          1
Genevieve      1
Name: name, Length: 957, dtype: int64

In [12]:
twitter_archive.rating_numerator.value_counts() 

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [13]:
#this tweet numerators need to be fixed
twitter_archive.query('rating_numerator == 75').text.str.extract('((?:\d+\.)?\d+)\/(\d+)', expand=True)

0   1
340  9.75  10
695  9.75  10

In [14]:
# numerators are not 75, 9.75.
twitter_archive.query('rating_numerator == 75')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
340  832215909146226688                    NaN                  NaN   
695  786709082849828864                    NaN                  NaN   

                     timestamp  \
340  2017-02-16 13:11:49 +0000   
695  2016-10-13 23:23:56 +0000   

                                                source  \
340  <a href="http://twitter.com/download/iphone" r...   
695  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
340  RT @dog_rates: This is Logan, the Chow who liv...         7.867091e+17   
695  This is Logan, the Chow who lived. He solemnly...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
340              4.196984e+09  2016-10-13 23:23:56 +0000   
695                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
340  https://twitter.com/dog_rates/status/786709082...                75   
695  https://twitter.com/dog_rates/status/786709082...                75   

     rating_denominator   name doggo floofer pupper puppo  
340                  10  Logan  None    None   None  None  
695                  10  Logan  None    None   None  None

In [15]:
# outliers
twitter_archive.query('rating_numerator == 1776')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
979  749981277374128128                    NaN                  NaN   

                     timestamp  \
979  2016-07-04 15:00:45 +0000   

                                                source  \
979  <a href="https://about.twitter.com/products/tw...   

                                                  text  retweeted_status_id  \
979  This is Atticus. He's quite simply America af....                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
979                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
979  https://twitter.com/dog_rates/status/749981277...              1776   

     rating_denominator     name doggo floofer pupper puppo  
979                  10  Atticus  None    None   None  None

In [16]:
# denominators should be 10
twitter_archive.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

quality:
- `tweet_id` datatype needs to be object (str)
- `timestamp` column is object, not datetime
- `timestamp` column contains +0000 which looks unnecessary
- retweet or reply tweets are unnecessary (only original dog ratings)
- `name` column has some strange names (a, an etc.)
- some `rating_numerator` has outliers
- `rating_numerator` of id 786709082849828864 is decimal
- some`rating_denominator` are not 10
- `doggo`, `floofer`, `pupper`, `puppo` columns (dog_stage column) have missing values, and some values are None instead of NaN

tidiness:
- `doggo`, `floofer` `pupper`, `puppo` column should be one column

## image_prediction

In [17]:
image_prediction.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [18]:
image_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


- some column headers are hard to understand(`img_num`, `p1`, `p1_conf`, `p1_dog`, etc.)
- `tweet_id` datatype needs to be object
- dog breed names in `p1`, `p2`, `p3` columns are inconsistent (lowercase, uppercase)

## tweepy_data

In [19]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
tweepy_data.head(1)

created_at                  id              id_str  \
0 2017-08-01 16:23:56+00:00  892420643555336193  892420643555336192   

                                           full_text  truncated  \
0  This is Phineas. He's a mystical boy. Only eve...      False   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   

   in_reply_to_status_id_str  in_reply_to_user_id  in_reply_to_user_id_str  \
0                        NaN                  NaN                      NaN   

  in_reply_to_screen_name                                               user  \
0                    None  {'id': 4196983835, 'id_str': '4196983835', 'na...   

   geo  coordinates place  contributors  is_quote_status  retweet_count  \
0  NaN          NaN  None           NaN            False           7703   

   favorite_count  favorited  retweeted  possibly_sensitive  \
0           36183      False      False                 0.0   

   possibly_sensitive_appealable lang retweeted_status  quoted_status_id  \
0                            0.0   en              NaN               NaN   

   quoted_status_id_str quoted_status_permalink quoted_status  
0                   NaN                     NaN           NaN

In [20]:
tweepy_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2331 non-null   datetime64[ns, UTC]
 1   id                             2331 non-null   int64              
 2   id_str                         2331 non-null   int64              
 3   full_text                      2331 non-null   object             
 4   truncated                      2331 non-null   bool               
 5   display_text_range             2331 non-null   object             
 6   entities                       2331 non-null   object             
 7   extended_entities              2059 non-null   object             
 8   source                         2331 non-null   object             
 9   in_reply_to_status_id          77 non-null     float64            
 10  in_reply_to_status_id_st

## Quality Issues

**twitter_archive** 

1- `tweet_id` datatype needs to be object (str)\
2- `timestamp` column is object, not datetime\
3- `timestamp` column contains +0000 which looks unnecessary\
4- retweet or reply tweets are unnecessary (only original dog ratings)\
5- `name` column has some strange names (a, an etc.)\
6- some `rating_numerator` has too high, or too lowscore\
7- `rating_numerator` of id 786709082849828864 is decimal\
8- some`rating_denominator` are not 10\
9- `doggo`, `floofer`, `pupper`, `puppo` columns (dog_stage column) have missing values, and some values are None instead of NaN

**image_prediction**

1- some column headers are hard to understand(`img_num`, `p1`, `p1_conf`, `p1_dog`, etc.)\
2- `tweet_id` datatype needs to be object\
3- dog breed names in `p1`, `p2`, `p3` columns are inconsistent (lowercase, uppercase)

**tweepy_data**

1- drop unneccesary columns (keep `id` , `retweet_count`, `favorite_count` columns)\
2- `id` column datatype needs to be object\
3- change id column name to `tweet_id`

## Tidiness

#### twitter_archive

1- `doggo`, `floofer`, `pupper`, `puppo` columns instead of a single column dog_stage
2- `rating_numerator`, `rating_denominator` columns should be one column(rating)

**After taking care of dirty and untidy data, all tables need to be merged.
According to the rule of tidy data:
1: each variables forms a column
2: each observation forms a row
3: each type of observational unit forms a table

And `twitter_archive`, `image_prediction`, `tweepy_data` are all about WeRateDog tweet information, hence it breaks the 3rd rule. I should merge all of them. 

# Data cleaning

Data cleaning has three steps.

**Define**: convert our assessments into defined cleaning tasks. These definitions also serve as an instruction list so others (or yourself in the future) can look at your work and reproduce it.

**Code**: convert those definitions to code and run that code.

**Test**: test your dataset, visually or with code, to make sure your cleaning operations worked.

In [21]:
# make a copy

In [22]:
twitter_archive_clean = twitter_archive.copy()
image_prediction_clean = image_prediction.copy()
tweepy_data_clean = tweepy_data.copy()

## twitter_archive_clean

### 1- `tweet_id` datatype needs to be object (str)\
### 2- `timestamp` column is object, not datetime\
### 3- `timestamp` column contains +0000 which looks unnecessary

#### Define

Change datatype from integer to string
Change datatype from string to datetime
Slice data

#### Code

In [23]:
# id to string
twitter_archive_clean.tweet_id = twitter_archive_clean.tweet_id.astype(str)
# remove +0000
twitter_archive_clean.timestamp = twitter_archive_clean.timestamp.str[:-6]
# timestamp to datetime
twitter_archive_clean.timestamp = pd.to_datetime(twitter_archive_clean.timestamp)

#### Test

In [24]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   tweet_id                    2356 non-null   object        
 1   in_reply_to_status_id       78 non-null     float64       
 2   in_reply_to_user_id         78 non-null     float64       
 3   timestamp                   2356 non-null   datetime64[ns]
 4   source                      2356 non-null   object        
 5   text                        2356 non-null   object        
 6   retweeted_status_id         181 non-null    float64       
 7   retweeted_status_user_id    181 non-null    float64       
 8   retweeted_status_timestamp  181 non-null    object        
 9   expanded_urls               2297 non-null   object        
 10  rating_numerator            2356 non-null   int64         
 11  rating_denominator          2356 non-null   int64       

In [25]:
twitter_archive_clean.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

            timestamp                                             source  \
0 2017-08-01 16:23:56  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

### 4- retweet or reply tweets are unnecessary (only original dog ratings)

#### Define

Drop retweets and replies, then drop unnecessary columns

#### Code

In [26]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype         
---  ------                      --------------  -----         
 0   tweet_id                    2356 non-null   object        
 1   in_reply_to_status_id       78 non-null     float64       
 2   in_reply_to_user_id         78 non-null     float64       
 3   timestamp                   2356 non-null   datetime64[ns]
 4   source                      2356 non-null   object        
 5   text                        2356 non-null   object        
 6   retweeted_status_id         181 non-null    float64       
 7   retweeted_status_user_id    181 non-null    float64       
 8   retweeted_status_timestamp  181 non-null    object        
 9   expanded_urls               2297 non-null   object        
 10  rating_numerator            2356 non-null   int64         
 11  rating_denominator          2356 non-null   int64       

In [27]:
# drop 78 replies
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.in_reply_to_status_id.isnull()]

In [28]:
# drop 181 retweets
twitter_archive_clean = twitter_archive_clean[twitter_archive_clean.retweeted_status_id.isnull()]

In [29]:
# 2356 - 78 - 181 = 2097
len(twitter_archive_clean)

2097

In [30]:
# drop unnecessary columns
coldrop = twitter_archive_clean.iloc[:,np.r_[1,2,6:9]].columns
twitter_archive_clean.drop(coldrop, axis = 1,inplace = True)

#### Test

In [31]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   tweet_id            2097 non-null   object        
 1   timestamp           2097 non-null   datetime64[ns]
 2   source              2097 non-null   object        
 3   text                2097 non-null   object        
 4   expanded_urls       2094 non-null   object        
 5   rating_numerator    2097 non-null   int64         
 6   rating_denominator  2097 non-null   int64         
 7   name                2097 non-null   object        
 8   doggo               2097 non-null   object        
 9   floofer             2097 non-null   object        
 10  pupper              2097 non-null   object        
 11  puppo               2097 non-null   object        
dtypes: datetime64[ns](1), int64(2), object(9)
memory usage: 213.0+ KB


### 5- name column has some strange names (a, an etc.)

#### Define

All dog names are capitalized, and some words extracted as names by mistake are all lowercase.
Make a filter to figure out all lowercase names and drop those names.

#### Code

In [32]:
# filter out all lowercase names
name_filter = twitter_archive_clean.name.str.islower()
column_name = 'name'
# drop lowercase names
twitter_archive_clean.loc[name_filter, column_name] = np.nan

#### Test

In [33]:
twitter_archive_clean.name.value_counts()

None              603
Lucy               11
Charlie            11
Oliver             10
Cooper             10
Tucker              9
Penny               9
Sadie               8
Winston             8
Lola                8
Toby                7
Daisy               7
Stanley             6
Bella               6
Koda                6
Bailey              6
Bo                  6
Jax                 6
Oscar               6
Dave                5
Rusty               5
Chester             5
Scout               5
Milo                5
Buddy               5
Louis               5
Bentley             5
Leo                 5
Cassie              4
Boomer              4
Derek               4
Reggie              4
Maggie              4
Winnie              4
Duke                4
Archie              4
Jack                4
Gary                4
Oakley              4
Alfie               4
Dexter              4
Sammy               4
Larry               4
Jeffrey             4
Phil                4
Sophie    

### 6- some rating_numerator has too high, or too lowscore
### 7- rating_numerator of id 786709082849828864 is decimal
### 8- some rating_denominator are not 10

#### Define

extract correct rating score from `text` column by using regular expressions, replace it to existing rating columns\
determine each tweet score by computing tweet score ratio\
drop outliers if possible

#### Code

In [34]:
twitter_archive_clean.text.str.extract('((?:\d+\.)?\d+)\/(\d+)', expand=True)

0    1
0        13   10
1        13   10
2        12   10
3        13   10
4        12   10
5        13   10
6        13   10
7        13   10
8        13   10
9        14   10
10       13   10
11       13   10
12       13   10
13       12   10
14       13   10
15       13   10
16       12   10
17       13   10
18       13   10
20       12   10
21       13   10
22       14   10
23       13   10
24       13   10
25       12   10
26       13   10
27       13   10
28       13   10
29       12   10
31       13   10
33       12   10
34       13   10
35       14   10
37       13   10
38       12   10
39       13   10
40       13   10
41       14   10
42       13   10
43       12   10
44       12   10
45     13.5   10
46       13   10
47       13   10
48       13   10
49       12   10
50       13   10
51       13   10
52       13   10
53       12   10
54       13   10
56       14   10
57       12   10
58       13   10
59       12   10
60       12   10
61       12   10
62       11   10
63       13   10
65       12   10
66       13   10
67       12   10
69       11   10
70       13   10
71       13   10
72       13   10
75       13   10
76       14   10
77       13   10
79       12   10
80       13   10
81       12   10
82       12   10
83       14   10
84       13   10
85       13   10
86       13   10
87       13   10
88       13   10
89       12   10
90       12   10
92       13   10
93       12   10
94       13   10
96       13   10
98       12   10
99       12   10
100      13   10
102      13   10
103      12   10
104      13   10
105      12   10
106      13   10
107      12   10
108      12   10
110      14   10
111      13   10
112      11   10
114      13   10
115      13   10
116      13   10
117      14   10
119      13   10
120      13   10
121      12   10
122      13   10
123      14   10
125      13   10
126      12   10
127      12   10
128      13   10
129      12   10
131      13   10
133      13   10
134      12   10
135      13   10
136      12   10
138      13   10
139      13   10
140      13   10
141      14   10
142      12   10
143      13   10
144      13   10
145      13   10
147      12   10
150      13   10
151      13   10
152      11   10
153      13   10
154      13   10
156      13   10
157      13   10
158      12   10
161      13   10
162      13   10
163      12   10
164      14   10
166      12   10
167      13   10
168      13   10
169      12   10
170      13   10
172      13   10
173      13   10
174      12   10
175      12   10
176      13   10
177      13   10
178      13   10
181      12   10
183      12   10
187      14   10
190      13   10
191      13   10
192      13   10
193      12   10
196      12   10
197      13   10
198      13   10
199      14   10
200      11   10
201      12   10
202      13   10
203      13   10
205      12   10
206      12   10
207      13   10
208      13   10
209      14   10
210      12   10
213      11   10
214      14   10
215      13   10
216      11   10
217      13   10
219      11   10
220      12   10
221      12   10
223      12   10
224      11   10
225      13   10
226      12   10
227      12   10
229       6   10
232      10   10
233      12   10
235      12   10
236      12   10
237      11   10
238      11   10
239      13   10
240      13   10
241      11   10
242      13   10
243      13   10
244      12   10
245      13   10
246      12   10
248      13   10
249      12   10
252      12   10
253      13   10
254      11   10
255      12   10
256      13   10
257      12   10
258      11   10
259      13   10
261      13   10
262      12   10
263      13   10
264      12   10
265      12   10
267      12   10
268      14   10
269      12   10
270      13   10
271      12   10
275      10   10
276      12   10
277      13   10
278      12   10
279      13   10
280      12   10
282      13   10
283      13   10
284      13   10
287      13   10
288      13   10
292      13   10
293      12   10
294      13   10
295      12   10
296      13   1

In [35]:
# get correct ratings
correct_rating = twitter_archive_clean.text.str.extract('((?:\d+\.)?\d+)\/(\d+)', expand=True)
twitter_archive_clean['rating_numerator'] = correct_rating
# some numerators are float
twitter_archive_clean['rating_numerator'] = twitter_archive_clean['rating_numerator'].astype(float)

In [36]:
# since denominators are not consistent, it is hard to determine which tweet has better score.
# So I'd like to compute numerator / denominator ratio to see tweets score.
twitter_archive_clean['rating_ratio'] = twitter_archive_clean.rating_numerator / twitter_archive_clean.rating_denominator

In [37]:
# there are some outliers, but it is much easier to see tweet score.
# I will not drop outliers since there are only 3 (0.00 , 177.60 , 42.00) , there should be no problem to do an analysis
twitter_archive_clean.rating_ratio.value_counts()

1.200000      490
1.000000      438
1.100000      418
1.300000      287
0.900000      153
0.800000       98
0.700000       51
1.400000       38
0.500000       34
0.600000       32
0.300000       19
0.400000       15
0.200000       10
0.100000        4
1.350000        1
3.428571        1
1.127000        1
0.636364        1
1.126000        1
177.600000      1
42.000000       1
0.975000        1
0.000000        1
0.818182        1
Name: rating_ratio, dtype: int64

In [38]:
# drop columns
twitter_archive_clean = twitter_archive_clean.drop(['rating_numerator'] , axis = 1)
twitter_archive_clean = twitter_archive_clean.drop(['rating_denominator'] , axis = 1)

In [39]:
twitter_archive_clean

tweet_id           timestamp  \
0     892420643555336193 2017-08-01 16:23:56   
1     892177421306343426 2017-08-01 00:17:27   
2     891815181378084864 2017-07-31 00:18:03   
3     891689557279858688 2017-07-30 15:58:51   
4     891327558926688256 2017-07-29 16:00:24   
5     891087950875897856 2017-07-29 00:08:17   
6     890971913173991426 2017-07-28 16:27:12   
7     890729181411237888 2017-07-28 00:22:40   
8     890609185150312448 2017-07-27 16:25:51   
9     890240255349198849 2017-07-26 15:59:51   
10    890006608113172480 2017-07-26 00:31:25   
11    889880896479866881 2017-07-25 16:11:53   
12    889665388333682689 2017-07-25 01:55:32   
13    889638837579907072 2017-07-25 00:10:02   
14    889531135344209921 2017-07-24 17:02:04   
15    889278841981685760 2017-07-24 00:19:32   
16    888917238123831296 2017-07-23 00:22:39   
17    888804989199671297 2017-07-22 16:56:37   
18    888554962724278272 2017-07-22 00:23:06   
20    888078434458587136 2017-07-20 16:49:33   
21    887705289381826560 2017-07-19 16:06:48   
22    887517139158093824 2017-07-19 03:39:09   
23    887473957103951883 2017-07-19 00:47:34   
24    887343217045368832 2017-07-18 16:08:03   
25    887101392804085760 2017-07-18 00:07:08   
26    886983233522544640 2017-07-17 16:17:36   
27    886736880519319552 2017-07-16 23:58:41   
28    886680336477933568 2017-07-16 20:14:00   
29    886366144734445568 2017-07-15 23:25:31   
31    886258384151887873 2017-07-15 16:17:19   
33    885984800019947520 2017-07-14 22:10:11   
34    885528943205470208 2017-07-13 15:58:47   
35    885518971528720385 2017-07-13 15:19:09   
37    885167619883638784 2017-07-12 16:03:00   
38    884925521741709313 2017-07-12 00:01:00   
39    884876753390489601 2017-07-11 20:47:12   
40    884562892145688576 2017-07-11 00:00:02   
41    884441805382717440 2017-07-10 15:58:53   
42    884247878851493888 2017-07-10 03:08:17   
43    884162670584377345 2017-07-09 21:29:42   
44    883838122936631299 2017-07-09 00:00:04   
45    883482846933004288 2017-07-08 00:28:19   
46    883360690899218434 2017-07-07 16:22:55   
47    883117836046086144 2017-07-07 00:17:54   
48    882992080364220416 2017-07-06 15:58:11   
49    882762694511734784 2017-07-06 00:46:41   
50    882627270321602560 2017-07-05 15:48:34   
51    882268110199369728 2017-07-04 16:01:23   
52    882045870035918850 2017-07-04 01:18:17   
53    881906580714921986 2017-07-03 16:04:48   
54    881666595344535552 2017-07-03 00:11:11   
56    881536004380872706 2017-07-02 15:32:16   
57    881268444196462592 2017-07-01 21:49:04   
58    880935762899988482 2017-06-30 23:47:07   
59    880872448815771648 2017-06-30 19:35:32   
60    880465832366813184 2017-06-29 16:39:47   
61    880221127280381952 2017-06-29 00:27:25   
62    880095782870896641 2017-06-28 16:09:20   
63    879862464715927552 2017-06-28 00:42:13   
65    879492040517615616 2017-06-27 00:10:17   
66    879415818425184262 2017-06-26 19:07:24   
67    879376492567855104 2017-06-26 16:31:08   
69    879050749262655488 2017-06-25 18:56:45   
70    879008229531029506 2017-06-25 16:07:47   
71    878776093423087618 2017-06-25 00:45:22   
72    878604707211726852 2017-06-24 13:24:20   
75    878281511006478336 2017-06-23 16:00:04   
76    878057613040115712 2017-06-23 01:10:23   
77    877736472329191424 2017-06-22 03:54:17   
79    877556246731214848 2017-06-21 15:58:08   
80    877316821321428993 2017-06-21 00:06:44   
81    877201837425926144 2017-06-20 16:29:50   
82    876838120628539392 2017-06-19 16:24:33   
83    876537666061221889 2017-06-18 20:30:39   
84    876484053909872640 2017-06-18 16:57:37   
85    876120275196170240 2017-06-17 16:52:05   
86    875747767867523072 2017-06-16 16:11:53   
87    875144289856114688 2017-06-15 00:13:52   
88    875097192612077568 2017-06-14 21:06:43   
89    875021211251597312 2017-06-14 16:04:48   
90    874680097055178752 2017-06-13 17:29:20   
92    874296783580663808 2017-06-12 16:06:11   
93    874057562936811520 2017-06

### 9- `doggo`, `floofer`, `pupper`, `puppo` columns (dog_stage column) have missing values, and some values are None instead of NaN

#### Define

Extract dog stage from text column\
Replace None to ' '\
Then put np.nan

#### Code

In [40]:
# doggo
twitter_archive_clean['doggo'] = twitter_archive_clean.text.str.extract('([Dd]oggo)')
twitter_archive_clean['doggo'].replace('None' , '' , inplace = True)
twitter_archive_clean.loc[twitter_archive_clean.doggo == "" , 'doggo'] = np.nan

In [41]:
# floofer
twitter_archive_clean['floofer'] = twitter_archive_clean.text.str.extract('([Ff]loofer)')
twitter_archive_clean['floofer'].replace('None' , '' , inplace = True)
twitter_archive_clean.loc[twitter_archive_clean.floofer == "" , 'floofer'] = np.nan

In [42]:
# pupper
twitter_archive_clean['pupper'] = twitter_archive_clean.text.str.extract('([Pp]upper)')
twitter_archive_clean['pupper'].replace('None' , '' , inplace = True)
twitter_archive_clean.loc[twitter_archive_clean.pupper == "" , 'pupper'] = np.nan

In [43]:
# puppo
twitter_archive_clean['puppo'] = twitter_archive_clean.text.str.extract('([Pp]uppo)')
twitter_archive_clean['puppo'].replace('None' , '' , inplace = True)
twitter_archive_clean.loc[twitter_archive_clean.puppo == "" , 'puppo'] = np.nan

#### Test

In [44]:
# there are still missing values
twitter_archive_clean.head(1)

tweet_id           timestamp  \
0  892420643555336193 2017-08-01 16:23:56   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   

                                       expanded_urls     name doggo floofer  \
0  https://twitter.com/dog_rates/status/892420643...  Phineas   NaN     NaN   

  pupper puppo  rating_ratio  
0    NaN   NaN           1.3

In [45]:
# although since the text column doesn't have dog stage information in some tweets, I have no way to get it.
print(twitter_archive_clean.at[0,'text'])

This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU


In [48]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   tweet_id       2097 non-null   object        
 1   timestamp      2097 non-null   datetime64[ns]
 2   source         2097 non-null   object        
 3   text           2097 non-null   object        
 4   expanded_urls  2094 non-null   object        
 5   name           1993 non-null   object        
 6   doggo          90 non-null     object        
 7   floofer        10 non-null     object        
 8   pupper         252 non-null    object        
 9   puppo          30 non-null     object        
 10  rating_ratio   2097 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(9)
memory usage: 276.6+ KB


In [46]:
# 1 - two column for tweet id( id and id_str)
import requests
id = 'https://twitter.com/anyuser/status/892420643555336193'
id_str = 'https://twitter.com/anyuser/status/892420643555336192'
response_id = requests.get(id)
response_id_str = requests.get(id_str)

print(response_id,response_id_str)

<Response [200]> <Response [200]>


In [47]:
# Change datatype od timestamp column to datetime
pass
twitter_archive['timestamp'] = pd.to_datetime(twitter_archive['timestamp'])